### ISSUE TITLE
 Constructor property promotion fixer

### ISSUE THREAD

**Issue by Bilge** [PHP-CS-Fixer/PHP-CS-Fixer#5956]
## Feature request

Convert simple class property assignments in constructor to constructor declarations.

```php
class Foo
{
    private bool $x;

    public function __construct(bool $x)
    {
        $this->x = $x;
    }
}
```

Becomes:

```php
class Foo
{
    public function __construct(
        private bool $x
    ) {}
}
```

**Comment by keradus** [comment]
hey @Bilge , interesting idea!
can you propose a PR for it, please?

**Comment by liquid207** [comment]
[rectorphp
/
rector](https://github.com/rectorphp/rector/blob/c7b2129f6435b6f822ac939697ebb7abc97f910a/rules/Php80/Rector/Class_/ClassPropertyAssignToConstructorPromotionRector.php#L33) already has this function, do PHP-CS-Fixer need to re-implement it?

**Comment by julienfalque** [comment]
While avoiding efforts duplication and favoring existing tools would be great, I think I would still like to have this feature here instead of installing another tool for it specifically.

**Comment by PHLAK** [comment]
I agree with @julienfalque, php-cs-fixer is intended to fix/enforce coding standards and the use of constructor property promotion is a coding standards choice. Therefore I believe it belongs in this project.

**Comment by Bilge** [comment]
@kubawerlos [already did it](https://github.com/kubawerlos/php-cs-fixer-custom-fixers/blob/a2cfe3118190f256a6e8761863c97dbac21c9251/src/Fixer/PromotedConstructorPropertyFixer.php). But the problem you have is [this](https://github.com/kubawerlos/php-cs-fixer-custom-fixers/issues/628).

**Comment by arnaud-lb** [comment]
> Convert simple class property assignments in constructor to constructor declarations.

Since promoted properties makes it faster to write code, but do not necessarily improve legibility, it would be nice if this rule also supported the conversion from promoted properties to plain properties. This way, it would be possible to write code with promoted properties and to get it converted to plain properties later to improve legibility.

**Comment by chu121su12** [comment]
Additional suggestion is to add configuration to, say, [`class_attributes_separation`](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/doc/rules/class_notation/class_attributes_separation.rst) to space properties for readability, since the `property` element do not cover these.

![image](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/379924/c639452b-a3d6-439b-b8a4-4346d414090e)

**Comment by Bilge** [comment]
It pains me to know there is someone in the world whose editor unironically uses a font like that.

**Comment by Wirone** [comment]
@chu121su12 adding empty line between attributed params was discussed in #7111 and should be a part of `method_argument_space`, not fixer proposed here. These two fixers should have integration test involving them, though.

In general: fixers should be as atomic as possible, with single responsibility. Then, they can be combined in bigger rule sets to achieve what's needed 🙂.

**Comment by maurei** [comment]
Waiting for this feature to automate a big part of cleanup in our codebase where we're looking to remove redundant code :)

>Since promoted properties makes it faster to write code, but do not necessarily improve legibility,

**Comment by Bilge** [comment]
Just use Rector. I'm unsubscribing from this thread now.

**Comment by keradus** [comment]
thus the labels
https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/5956#event-9542498633
